src/workerd/api/streams/standard.c++

```diff
@@ -2080,382 +2072,361 @@ ReadableStreamJsController::getController() {
-      state(kj::mv(consumer)) {
-  state.get<kj::Own<ValueReadable>>()->setOwner(this);
-}
+  void doClose() override {
+    state.template init<StreamStates::Closed>();
+  }
 
-ReadableStreamJsSource::ReadableStreamJsSource(kj::Own<ByteReadable> consumer)
-    : ioContext(IoContext::current()),
-      state(kj::mv(consumer)) {
-  state.get<kj::Own<ByteReadable>>()->setOwner(this);
-}
+  void doError(jsg::Lock& js, v8::Local<v8::Value> reason) override {
+    state.template init<StreamStates::Errored>(js.v8Ref(reason));
+  }
 
-void ReadableStreamJsSource::cancel(kj::Exception reason) {
-  const auto doCancel = [this](auto& consumer, auto reason) {
-    auto c = kj::mv(consumer);
-    state.init<kj::Exception>(kj::cp(reason));
-    ioContext.addTask(ioContext.run(
-        [consumer = kj::mv(c), reason = kj::mv(reason)]
-        (Worker::Lock& lock) mutable -> kj::Promise<void> {
-      jsg::Lock& js = lock;
-      v8::HandleScope handleScope(js.v8Isolate);
-      return IoContext::current().awaitJs(
-          consumer->cancel(js, js.exceptionToJs(kj::cp(reason)).getHandle(js)));
-    }).attach(kj::addRef(*this)));
-  };
+  jsg::Promise<kj::Array<byte>> allBytes(jsg::Lock& js) {
+    return loop(js).then(js, [this](auto& js, PartList&& partPtrs) {
+      auto out = kj::heapArray<byte>(runningTotal);
+      copyInto(out, kj::mv(partPtrs));
+      return kj::mv(out);
+    });
+  }
 
-  KJ_SWITCH_ONEOF(state) {
-    KJ_CASE_ONEOF(closed, StreamStates::Closed) { return; }
-    KJ_CASE_ONEOF(errored, kj::Exception) { kj::throwFatalException(kj::cp(errored)); }
-    KJ_CASE_ONEOF(consumer, kj::Own<ValueReadable>) {
-      return doCancel(consumer, kj::mv(reason));
-    }
-    KJ_CASE_ONEOF(consumer, kj::Own<ByteReadable>) {
-      return doCancel(consumer, kj::mv(reason));
-    }
+  jsg::Promise<kj::String> allText(jsg::Lock& js) {
+    return loop(js).then(js, [this](auto& js, PartList&& partPtrs) {
+      auto out = kj::heapArray<char>(runningTotal + 1);
+      copyInto(out.slice(0, out.size() - 1).asBytes(), kj::mv(partPtrs));
+      out.back() = '\0';
+      return kj::String(kj::mv(out));
+    });
   }
-  KJ_UNREACHABLE;
-}
 
-void ReadableStreamJsSource::doClose() {
-  if (readPending) {
-    // If we're closed in the middle of a read, we do not want to actually
-    // change the state yet we still have data to receive and process. Instead,
-    // we'll set a flag indicating that we are closing so that once the read
-    // completes, we'll process the close after all data is drained.
-    closePending = true;
-    return;
+  kj::Promise<void> pumpTo(jsg::Lock& js, kj::Own<WritableStreamSink> sink, bool end) {
+    auto& ioContext = IoContext::current();
+    return ioContext.awaitJs(pumpLoop(js,
+        kj::refcounted<RefcountedSink>(ioContext.addObject(kj::mv(sink))),
+        end));
   }
-  state.init<StreamStates::Closed>();
-}
+private:
+  struct RefcountedSink: public WritableStreamSink, public kj::Refcounted {
```

Comment:
Keep in mind that since the refcounting is not thread-safe, thus it will be critical that anywhere where you use `Own<RefcountedSink>` (or `Own<WritableStreamSink>` pointing at a `RefcountedSink`) must be protected by the isolate lock. This means in particular you cannot attach `Own<RefcountedSink>` to a KJ promise, because the promise could be canceled while outside of the isolate lock, which would decrement the refcount without the lock held.

I haven't checked the code in this PR to see if it's safe, but this kind of gotcha makes me nervous either way.

Reply:
Yeah, wasn't really happy with this bit.  I'll take another pass at these bits.

Reply:
Alright, took a stab at reworking this. Take a look at the last fixup commit. @kentonv ... I'd forgotten that `ioContext.addFunctor(...)` existed, thank you for the reminder. Unfortunately I'm not entirely sure if I'm using it correctly here and the other couple of examples and docs aren't really all that clear.